File: data/application/appsflyer-onelink-ios-sample-apps/obj-c/obj-c/DLViewController.m
#import "DLViewController.h"

@interface DLViewController ()
- (void)copyShareInviteLinkWithFruitName:(NSString *)fruitName;
@end

@implementation DLViewController

- (void)viewDidLoad {
    [super viewDidLoad];
}

- (NSMutableAttributedString *)attributionDataToStringWithData:(NSDictionary<NSString *, id> *)data {
    NSMutableAttributedString *newString = [[NSMutableAttributedString alloc] init];
    NSDictionary *boldAttribute = @{NSFontAttributeName: [UIFont fontWithName:@"Avenir Next Bold" size:18.0]};
    NSDictionary *regularAttribute = @{NSFontAttributeName: [UIFont fontWithName:@"Avenir Next" size:18.0]};

    NSArray<NSString *> *sortedKeys = [data.allKeys sortedArrayUsingSelector:@selector(compare:)];

    for (NSString *key in sortedKeys) {
        NSLog(@"ViewController %@ : %@", key, data[key] ?: @"null");

        NSAttributedString *boldKeyStr = [[NSAttributedString alloc] initWithString:key attributes:boldAttribute];
        [newString appendAttributedString:boldKeyStr];

        NSString *valueStr;

        if ([data[key] isKindOfClass:[NSString class]]) {
            valueStr = data[key];
        } else if ([data[key] isKindOfClass:[NSNumber class]]) {
            valueStr = [(NSNumber *)data[key] stringValue];
        } else {
            valueStr = @"null";
        }

        NSAttributedString *normalValueStr = [[NSAttributedString alloc] initWithString:[NSString stringWithFormat:@": %@\n", valueStr] attributes:regularAttribute];
        [newString appendAttributedString:normalValueStr];
    }

    return newString;
}

- (void)showToastWithMessage:(NSString *)message font:(UIFont *)font {
    UILabel *toastLabel = [[UILabel alloc] initWithFrame:CGRectMake(self.view.frame.size.width / 2 - 75, 20, 150, 35)];
    toastLabel.backgroundColor = [[UIColor blackColor] colorWithAlphaComponent:0.6];
    toastLabel.textColor = [UIColor whiteColor];
    toastLabel.font = font;
    toastLabel.textAlignment = NSTextAlignmentCenter;
    toastLabel.text = message;
    toastLabel.alpha = 1.0;
    toastLabel.layer.cornerRadius = 10;
    toastLabel.clipsToBounds = YES;
    [self.view addSubview:toastLabel];

    [UIView animateWithDuration:4.0 delay:0.1 options:UIViewAnimationOptionCurveEaseOut animations:^{
        toastLabel.alpha = 0.0;
    } completion:^(BOOL isCompleted) {
        [toastLabel removeFromSuperview];
    }];
}

- (void)copyShareInviteLinkWithFruitName:(NSString *)fruitName {
    [self showToastWithMessage:@"Share invite requires AppsFlyer SDK" font:[UIFont systemFontOfSize:12.0]];
}

- (NSString *)getFruitAmountWithData:(NSDictionary<NSString *, id> *)data {
    NSSet<NSString *> *keys = [NSSet setWithArray:data.allKeys];
    id fruitAmount = nil;

    if ([keys containsObject:@"deep_link_value"] && [keys containsObject:@"deep_link_sub1"]) {
        fruitAmount = data[@"deep_link_sub1"];
        NSLog(@"deep_link_sub1 found and is %@", fruitAmount);
    } else if ([keys containsObject:@"fruit_name"] && [keys containsObject:@"fruit_amount"]) {
        fruitAmount = data[@"fruit_amount"];
        NSLog(@"fruit_amount found and is %@", fruitAmount);
    } else {
        NSLog(@"deep_link_sub1/fruit_amount not found");
        return nil;
    }

    NSCharacterSet *decimalDigits = [NSCharacterSet decimalDigitCharacterSet];
    if ([fruitAmount isKindOfClass:[NSString class]] && [decimalDigits isSupersetOfSet:[NSCharacterSet characterSetWithCharactersInString:(NSString *)fruitAmount]]) {
        self.fruitAmountStr = (NSString *)fruitAmount;
        return (NSString *)fruitAmount;
    } else {
        NSLog(@"Fruit amount is not a whole number");
        return nil;
    }
}

@end

File: data/application/appsflyer-onelink-ios-sample-apps/swift/basic_app/basic_app/DLViewController.swift
//
//  DLViewController.swift
//  basic_app
//
//  Created by Liaz Kamper on 01/06/2020.
//  Copyright © 2020 OneLink. All rights reserved.
//

import UIKit

class DLViewController: UIViewController {

    var deepLinkData: [String: Any]? = nil
    var fruitAmountStr: String = "000"

    override func viewDidLoad() {
        super.viewDidLoad()
    }

    func attributionDataToString(data : [String: Any]) -> NSMutableAttributedString {
        let newString = NSMutableAttributedString()
        let boldAttribute = [
           NSAttributedString.Key.font: UIFont(name: "Avenir Next Bold", size: 18.0)!
        ]
        let regularAttribute = [
           NSAttributedString.Key.font: UIFont(name: "Avenir Next", size: 18.0)!
        ]
        let sortedKeys = Array(data.keys).sorted(by: <)
        for key in sortedKeys {
            print("ViewController", key, ":",data[key] ?? "null")
            let keyStr = key
            let boldKeyStr = NSAttributedString(string: keyStr, attributes: boldAttribute)
            newString.append(boldKeyStr)

            var valueStr: String
            switch data[key] {
            case let s as String:
                valueStr = s
            case let b as Bool:
                valueStr = b.description
            default:
                valueStr = "null"
            }

            let normalValueStr = NSAttributedString(string: ": \(valueStr)\n", attributes: regularAttribute)
            newString.append(normalValueStr)
        }
        return newString
    }

    func showToast(message : String, font: UIFont) {
        let toastLabel = UILabel(frame: CGRect(x: self.view.frame.size.width/2 - 75, y: 20, width: 150, height: 35))
        toastLabel.backgroundColor = UIColor.black.withAlphaComponent(0.6)
        toastLabel.textColor = UIColor.white
        toastLabel.font = font
        toastLabel.textAlignment = .center;
        toastLabel.text = message
        toastLabel.alpha = 1.0
        toastLabel.layer.cornerRadius = 10;
        toastLabel.clipsToBounds  =  true
        self.view.addSubview(toastLabel)
        UIView.animate(withDuration: 4.0, delay: 0.1, options: .curveEaseOut, animations: {
             toastLabel.alpha = 0.0
        }, completion: {(isCompleted) in
            toastLabel.removeFromSuperview()
        })
    }

    func copyShareInviteLink(fruitName: String){
        showToast(message: "Share invite requires AppsFlyer SDK", font: .systemFont(ofSize: 12.0))
    }

    func getFruitAmount(data : [String: Any]) -> String? {
        let keys = data.keys
        var fruitAmount: Any?
        if keys.contains("deep_link_value") && keys.contains("deep_link_sub1"){
            fruitAmount = data["deep_link_sub1"]
            NSLog("deep_link_sub1 found and is \(fruitAmount!)")
        }
        else if keys.contains("fruit_name") && keys.contains("fruit_amount"){
            fruitAmount = data["fruit_amount"]
            NSLog("fruit_amount found and is \(fruitAmount!)")
        }
        else {
            NSLog("deep_link_sub1/fruit_amount not found")
            return nil
        }

        guard CharacterSet.decimalDigits.isSuperset(of: CharacterSet(charactersIn: fruitAmount as! String)) else {
            NSLog("Fruit amount is not a whole number")
            return nil
        }
        fruitAmountStr = fruitAmount as? String ?? "000"
        return fruitAmount as? String
    }
}
